test/membrane/error.spec.js

```diff
@@ -0,0 +1,212 @@
+                })
+            `)
+})
+
+it('unhandled promise rejections with non-error objects', (done) => {
+    const evalScript = createSecureEnvironment(undefined, { done, expect })
+    evalScript(`
+        const errorObj = { foo: 'bar' }
+        window.addEventListener("unhandledrejection", event => {
+            expect(event.reason).toBe(errorObj)
+            expect(event.reason.foo).toBe('bar')
+            done()
+        });
+
+        new Promise((resolve, reject) => {
+            throw errorObj
+        });
+    `)
+})
+
+it('non-error objects thrown in blue functions do not leak', () => {
+    function foo() {
+        throw { foo: 'bar' }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { foo, expect })
+    evalScript(`
+                try {
+                    foo()
+                } catch (e) {
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe(undefined)
+                }
+            `)
+})
+
+it('non-error objects thrown in blue constructors do not leak', () => {
+    class Foo {
+        constructor() {
+            throw { foo: 'bar' }
+        }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { Foo, expect })
+    evalScript(`
+                try {
+                    new Foo()
+                } catch (e) {
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe(undefined)
+                }
+            `)
+})
+
+it('non-error objects thrown in red functions do not leak', () => {
+    let fn
+    function save(arg) {
+        fn = arg
+    }
+
+    const evalScript = createSecureEnvironment(undefined, { save })
+    evalScript(`
+        function foo() {
+            throw { foo: 'bar' }
+        }
+    `)
+
+    try {
+        fn()
+    } catch(e) {
+        expect(e.foo).toBe(undefined);
+    }
+})
+
+it('non-error objects thrown in red consturctors do not leak', () => {
+    let ctor
+    function save(arg) {
+        ctor = arg
+    }
+
+    const evalScript = createSecureEnvironment(undefined, { save })
+    evalScript(`
+        class Foo {
+            constructor() {
+                throw { foo: 'bar' }
+            }
+        }
+    `)
+
+    try {
+        new ctor()
+    } catch(e) {
+        expect(e.foo).toBe(undefined)
+    }
+})
+
+it('blue extended error objects preserve properties', () => {
+    class CustomError extends Error {
+        constructor(message) {
+            super(message)
+            this.bar = 'baz'
+        }
+
+        get foo() {
+            return 'bar'
+        }
+    }
+
+    class Foo {
+        constructor() {
+            throw new CustomError('foo');
+        }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { Foo, expect })
+    evalScript(`
+                try {
+                    new Foo()
+                } catch (e) {
+                    // blue error own properties do not leak
```

Comment:
not sure I understand this comment... but the test is fine!

Reply:
+1